### ISSUE TITLE
jax tutorial 4 (init/opt): fails to init model on GPU

### ISSUE THREAD

**Issue by murphyk** [phlippe/uvadlc_notebooks#83]
**Tutorial**: -1 (Fill-in number of tutorial)

https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/JAX/tutorial4/Optimization_and_Initialization.html#Constant-initialization

**Describe the bug**

This line fails
```
model, params = init_simple_model(get_const_init_func(c=0.005))
```
producing
```
--------------------------------------------------------------------------
XlaRuntimeError                           Traceback (most recent call last)
[<ipython-input-14-eb8cfd8e667a>](https://localhost:8080/#) in <module>
      7 
      8 #model, params = init_simple_model(get_const_init_func(c=0.005))
----> 9 model, params = init_simple_model(nn.linear.default_kernel_init)
     10 visualize_gradients(model, params)
     11 visualize_activations(model, params, print_variance=True)

24 frames
[<ipython-input-10-072d6886ae7a>](https://localhost:8080/#) in init_simple_model(kernel_init, act_fn)
      2     model = BaseNetwork(act_fn=act_fn,
      3                         kernel_init=kernel_init)
----> 4     params = model.init(random.PRNGKey(42), exmp_imgs)
      5     return model, params

[/usr/local/lib/python3.9/dist-packages/jax/_src/random.py](https://localhost:8080/#) in PRNGKey(seed)
    134     raise TypeError("PRNGKey accepts a scalar seed, but was given an array of"
    135                     f"shape {np.shape(seed)} != (). Use jax.vmap for batching")
--> 136   key = prng.seed_with_impl(impl, seed)
    137   return _return_prng_keys(True, key)

...

[/usr/local/lib/python3.9/dist-packages/jax/_src/dispatch.py](https://localhost:8080/#) in backend_compile(backend, built_c, options, host_callbacks)
   1034   # TODO(sharadmv): remove this fallback when all backends allow `compile`
   1035   # to take in `host_callbacks`
-> 1036   return backend.compile(built_c, compile_options=options)
   1037 
   1038 _ir_dump_counter = itertools.count()

XlaRuntimeError: INTERNAL: RET_CHECK failure (external/org_tensorflow/tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:641) dnn != nullptr
```

The problem also would occur in https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/JAX/tutorial3/Activation_Functions.html 
if the training was actually triggered, since it is the same model and data.

**To Reproduce (if any steps necessary)**

To try to isolate the problem, I created this[ minimal colab](https://colab.research.google.com/drive/1E5j2b8hIIxFtxiEdwQN8MOTKf2ZgXtNT?usp=sharing). 
This just tries to initialize the network using a batch of MNIST images.
It works on CPU but not GPU.

**Comment by murphyk** [comment]
The problem is triggered merely by using `key = jax.random.PRNGKey(0)`. I think it's because you mix in pytorch randomness in the dataloader....

**Comment by murphyk** [comment]
So I verified that if you access the jax RNG before the pytorch RNG, it works.
eg try running this before the dataloader is created
```

act_fn = lambda x: x #act_fn_by_name['identity']
model = BaseNetwork(act_fn=act_fn)
key = jax.random.PRNGKey(0) 
X = jax.random.normal(key, (1000, 28, 28))
params = model.init(random.PRNGKey(42), X) # not image data 
```

**Comment by murphyk** [comment]
If I put this at the top of the colab then tutorial 4 works
```
key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (10,10)) # force jax PRNG to run on GPU before pytorch grabs it
```

**Comment by phlippe** [comment]
Hi, thanks for raising the issue, I have not seen that behavior before. I have fixed it for the tutorials in efbc9de8a08857f1d074d1063c8c3c861475eea4, but I think it would be great if you can raise this issue to the JAX community as well, since it seems to be a general problem when wanting to use PyTorch's data loading functionality in JAX.

**Comment by murphyk** [comment]
I created https://github.com/google/jax/issues/15084.
